src/system/file_path.cc

```diff
@@ -16,123 +16,110 @@
     }
 
-    substitute_value();
+    substitute_value(file_path);
+	abs_file_path_ = boost::filesystem::path(file_path);
     if (ft == input_file) {
-    	if ( FilePath::is_absolute_path(abs_file_path_) ) {
+    	if ( abs_file_path_.is_absolute() ) {
     	} else {
-            abs_file_path_ = root_dir + DIR_DELIMITER + abs_file_path_;
+            abs_file_path_ = root_dir / file_path;
     	}
     } else if (ft == output_file) {
-        if ( FilePath::is_absolute_path(abs_file_path_) ) {
-            if (abs_file_path_.substr(0, output_dir.size()) == output_dir) {
-            	abs_file_path_=abs_file_path_.substr(output_dir.size()+1);
+        if ( abs_file_path_.is_absolute() ) {
+            if (abs_file_path_.string().substr(0, output_dir.string().size()) == output_dir.string()) {
+            	abs_file_path_ = boost::filesystem::path( abs_file_path_.string().substr(output_dir.string().size()+1) );
             } else {
-                THROW( ExcAbsOutputPath() << EI_Path( abs_file_path_ ) );
+                THROW( ExcAbsOutputPath() << EI_Path( abs_file_path_.string() ) );
             }
         }
-        abs_file_path_ = output_dir + DIR_DELIMITER + abs_file_path_;
+        abs_file_path_ = output_dir / abs_file_path_;
     }
 
 }
 
 
 
-void FilePath::set_io_dirs(const string working_dir, const string root_input_dir,const string input,const string output) {
+void FilePath::set_io_dirs(const string working_dir, const string root, const string input, const string output) {
+	ASSERT_EQ(working_dir, ".").error();
+    FilePath::set_dirs(root, input, output);
+}
+
+
+void FilePath::set_dirs(const string root, const string input, const string output) {
     // root directory
-    root_dir = root_input_dir;
-
-    // relative output dir is relative to working directory
-    // this is possibly independent of position of the main input file
-	output_dir = "";
-    if ( FilePath::is_absolute_path(output) ) {
-    	vector<string> dirs;
-    	boost::split(dirs, output,  boost::is_any_of("/"));
-    	for (vector<string>::iterator it = dirs.begin(); it != dirs.end(); ++it) {
-    	    if ( !(*it).size() ) continue;
-    	    if ( !output_dir.size() ) {
-#ifdef FLOW123D_HAVE_CYGWIN
-    	    	output_dir = (*it);
-#else
-    	    	output_dir = DIR_DELIMITER + *it;
-#endif // FLOW123D_HAVE_CYGWIN
-    	    } else {
-            	output_dir = output_dir + DIR_DELIMITER + *it;
-    	    }
-    	    FilePath::create_dir(output_dir);
-        }
+	root_dir = boost::filesystem::path(root);
+
+	// set output directory
+	// if param output is absolute output_dir = output
+	// else output_dir = root / output
+	// the resulting relative path is always completed to absulute path
+	boost::filesystem::path output_path = boost::filesystem::path(output);
+    if ( !output_path.is_absolute() ) {
+    	boost::filesystem::path root_output_path = boost::filesystem::path(root) / output_path;
+    	boost::filesystem::create_directories(root_output_path);
```

Comment:
Fine., just 
1. set e.g full_output_path for all cases
2. call create_directories for full_output_path
3. set output_dir to canonical (full_output_path)

The point is to call both `create_dirs` and `canonical` for all cases.